File: sdk/alcedocore-sdk-python/alcedo_sdk/resources/collections.py
from typing import Any, Optional

from .base import BaseResource


class CollectionsResource(BaseResource):
    async def list(self) -> dict:
        return await self._request("GET", "/api/collections")

    async def get(self, name: str) -> dict:
        return await self._request("GET", f"/api/collections/{name}")

    async def create(self, name: str, data: dict) -> dict:
        return await self._request("POST", f"/api/collections/{name}", json=data)

    async def update(self, name: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/collections/{name}", json=data)

    async def delete(self, name: str) -> dict:
        return await self._request("DELETE", f"/api/collections/{name}")

    async def list_sections(self, name: str) -> dict:
        return await self._request("GET", f"/api/collections/{name}/sections")

    async def create_section(self, name: str, data: dict) -> dict:
        return await self._request("POST", f"/api/collections/{name}/sections", json=data)

    async def update_section(self, name: str, section_id: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/collections/{name}/sections/{section_id}", json=data)

    async def delete_section(self, name: str, section_id: str) -> dict:
        return await self._request("DELETE", f"/api/collections/{name}/sections/{section_id}")

    async def list_views(self, name: str) -> dict:
        return await self._request("GET", f"/api/collections/{name}/views")

    async def create_view(self, name: str, data: dict) -> dict:
        return await self._request("POST", f"/api/collections/{name}/views", json=data)

    async def update_view(self, name: str, view_id: str, data: dict) -> dict:
        return await self._request("PUT", f"/api/collections/{name}/views/{view_id}", json=data)

    async def delete_view(self, name: str, view_id: str) -> dict:
        return await self._request("DELETE", f"/api/collections/{name}/views/{view_id}")

    async def set_default_view(self, name: str, view_id: str) -> dict:
        return await self._request("PUT", f"/api/collections/{name}/views/{view_id}/default")

File: sdk/alcedocore-sdk-node/src/collections.ts
import { KyInstance } from "ky";

export interface CollectionField {
    default?: any;
    display_name: string;
    name: string;
    type: string;
    required: boolean;
    unique: boolean;
    is_system: boolean;
}

export interface Collection {
    created_at: string;
    display_name?: string;
    fields: CollectionField[];
    is_system: boolean;
    name: string;
    updated_at: string;
}

export function createCollectionsResource(ky: KyInstance) {
    return {
        list: (options?: any) =>
            ky
                .get("collections", options)
                .json<{ collections: Collection[] }>(),
        get: (name: string, options?: any) =>
            ky
                .get(`collections/${encodeURIComponent(name)}`, options)
                .json<Collection>(),
        create: (data: any, options?: any) =>
            ky
                .post("collections", { json: data, ...options })
                .json<Collection>(),
        update: (name: string, data: any, options?: any) =>
            ky
                .put(`collections/${encodeURIComponent(name)}`, {
                    json: data,
                    ...options,
                })
                .json<{ success: true }>(),
        delete: (name: string, options?: any) =>
            ky
                .delete(`collections/${encodeURIComponent(name)}`, options)
                .json<{ deleted: true }>(),

        // Layouts
        listLayouts: (name: string, options?: any) =>
            ky
                .get(`collections/${encodeURIComponent(name)}/layouts`, options)
                .json(),
        createLayout: (name: string, data: any, options?: any) =>
            ky
                .post(`collections/${encodeURIComponent(name)}/layouts`, {
                    json: data,
                    ...options,
                })
                .json(),
        updateLayout: (
            name: string,
            layoutId: string,
            data: any,
            options?: any,
        ) =>
            ky
                .put(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}`,
                    { json: data, ...options },
                )
                .json(),
        deleteLayout: (name: string, layoutId: string, options?: any) =>
            ky
                .delete(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}`,
                    options,
                )
                .json(),
        getLayoutRoles: (name: string, layoutId: string, options?: any) =>
            ky
                .get(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/roles`,
                    options,
                )
                .json(),
        setLayoutRoles: (
            name: string,
            layoutId: string,
            roleIds: string[],
            options?: any,
        ) =>
            ky
                .put(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/roles`,
                    { json: { role_ids: roleIds }, ...options },
                )
                .json(),
        getResolvedLayout: (name: string, options?: any) =>
            ky
                .get(`collections/${encodeURIComponent(name)}/layout`, options)
                .json(),

        // Layout-scoped sections
        listLayoutSections: (name: string, layoutId: string, options?: any) =>
            ky
                .get(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/sections`,
                    options,
                )
                .json(),
        createLayoutSection: (
            name: string,
            layoutId: string,
            data: any,
            options?: any,
        ) =>
            ky
                .post(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/sections`,
                    { json: data, ...options },
                )
                .json(),
        updateLayoutSection: (
            name: string,
            layoutId: string,
            sectionId: string,
            data: any,
            options?: any,
        ) =>
            ky
                .put(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/sections/${encodeURIComponent(sectionId)}`,
                    { json: data, ...options },
                )
                .json(),
        deleteLayoutSection: (
            name: string,
            layoutId: string,
            sectionId: string,
            options?: any,
        ) =>
            ky
                .delete(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/sections/${encodeURIComponent(sectionId)}`,
                    options,
                )
                .json(),
        batchReorderSections: (
            name: string,
            layoutId: string,
            sections: any[],
            options?: any,
        ) =>
            ky
                .patch(
                    `collections/${encodeURIComponent(name)}/layouts/${encodeURIComponent(layoutId)}/sections`,
                    { json: { sections }, ...options },
                )
                .json(),

        // Saved views
        listViews: (name: string, options?: any) =>
            ky
                .get(`collections/${encodeURIComponent(name)}/views`, options)
                .json(),
        createView: (name: string, data: any, options?: any) =>
            ky
                .post(`collections/${encodeURIComponent(name)}/views`, {
                    json: data,
                    ...options,
                })
                .json(),
        updateView: (name: string, viewId: string, data: any, options?: any) =>
            ky
                .put(
                    `collections/${encodeURIComponent(name)}/views/${encodeURIComponent(viewId)}`,
                    { json: data, ...options },
                )
                .json(),
        deleteView: (name: string, viewId: string, options?: any) =>
            ky
                .delete(
                    `collections/${encodeURIComponent(name)}/views/${encodeURIComponent(viewId)}`,
                    options,
                )
                .json(),
        setDefaultView: (name: string, viewId: string, options?: any) =>
            ky
                .put(
                    `collections/${encodeURIComponent(name)}/views/${encodeURIComponent(viewId)}/default`,
                    options,
                )
                .json(),
        getCreatePolicy: (name: string, options?: any) =>
            ky
                .get(`collections/${encodeURIComponent(name)}/$create`, options)
                .json(),
    };
}
